torch/csrc/jit/codegen/cuda/disjoint_set.h

```diff
@@ -16,203 +16,241 @@ namespace fuser {
-    int num_sets = 0;
-    for (const auto& kv : entry_map) {
-      int fixed_point = fixedPoint(kv.first);
-      num_sets = std::max(num_sets, fixed_point + 1);
-      auto it = fixed_point_map.find(fixed_point);
-      if (it == fixed_point_map.end()) {
-        it = fixed_point_map.insert({fixed_point, {}}).first;
+  std::string toString() {
+    std::stringstream ss;
+    ss << "{ ";
+    for (auto entry : vector()) {
+      ss << abstractToString(entry);
+      if (entry != vector().back()) {
+        ss << "; ";
       }
-      it->second.insert(kv.first);
     }
+    ss << " }";
+    return ss.str();
+  }
 
-    std::vector<std::unordered_set<T, Hash>> disjoint_sets;
-    disjoint_sets.resize(num_sets);
-    for (auto entry : fixed_point_map) {
-      TORCH_INTERNAL_ASSERT(
-          entry.first < num_sets, "Unexpected error generating disjoint sets.");
-      disjoint_sets[entry.first] = entry.second;
-    }
+ private:
+  std::vector<T> vector_;
+  std::unordered_set<T, Hash> set_;
+};
 
-    // Vector may contain some empty sets, remove them
-    disjoint_sets.erase(
-        std::remove_if(
-            disjoint_sets.begin(),
-            disjoint_sets.end(),
-            [](std::unordered_set<T, Hash> set) { return set.empty(); }),
-        disjoint_sets.end());
+//! Container class DisjointSet models equivalence relationships
+//!
+//! Each instance of this class keeps equivalence sets
+//! DisjointSet::mapEntries(a,b) makes the full set of a and b equivalent
+//! DisjointSet::*AreMapped(a,b) checks if a and b belong to the same disjoint
+//! set
+template <typename T, typename Hash = std::hash<T>>
+class DisjointSets {
+ public:
+  DisjointSets() = default;
 
-    return disjoint_sets;
+  // Warning: returned values should never be modified. This accessor isn't
+  // strictly safe as VectorOfUniqueEntries is not returned as a const.
+  const std::
+      unordered_map<T, std::shared_ptr<VectorOfUniqueEntries<T, Hash>>, Hash>&
+      disjointSetMap() const {
+    return disjoint_set_maps_;
   }
 
-  //! Dumps the equivalent relationships
-  // TOOD: Change to "toString" for consistency
-  std::ostream& print(std::ostream& os) const {
-    std::unordered_map<int, std::unordered_set<T, Hash>> fixed_point_map;
-    for (const auto& kv : entry_map) {
-      int fixed_point = fixedPoint(kv.first);
-      auto it = fixed_point_map.find(fixed_point);
-      if (it == fixed_point_map.end()) {
-        it = fixed_point_map.insert({fixed_point, {}}).first;
-      }
-      it->second.insert(kv.first);
+  // Warning: returned values should never be modified. This accessor isn't
+  // strictly safe as VectorOfUniqueEntries is not returned as a const.
+  const std::vector<std::shared_ptr<VectorOfUniqueEntries<T, Hash>>>&
+  disjointSets() const {
+    return disjoint_sets_;
+  }
+
+  // Return the entire disjoint set of provided entry
+  const VectorOfUniqueEntries<T, Hash>& getDisjointSetOf(T entry) const {
```

Comment:
This class is most likely just used with a pointer type, so it doesn't matter, but general speaking, the parameter type should be `const T&`.